### DISCUSSION TITLE
Cumulative percentage per customer

### DISCUSSION THREAD

**Discussion by Lomarques-s** [DaxStudio/DaxStudio#1033] [General]
Hello DAX masters

I'm trying to create a measure that calculates the cumulative percentage per customer so that the total equals 100%.

I've seen several similar issues on the internet, but I haven't found anything that solves my problem. Note, I'm using Power Pivot.

I'm trying to create a measure using DAX Stidio, but I don't know why my measure doesn't work, the cumulative percentage always results in the number 1.00001 and does not bring the cumulative sum.

See in the print my frustrated attempt to measure

![image](https://user-images.githubusercontent.com/121581060/213548737-6ff34faa-9665-487c-9131-5492ad7e0ac7.png)

**Comment by dgosbell** [comment]
The problem here is that your Share_Contexto variable will be calculated once at the start of the query and then will use the same amount for all rows (which is why you get the same percentage for each row.

You need to either move the expression inline inside the FILTER() call and wrap it in CALCULATE() or declare it as a measure.

**Reply by Lomarques-s** [reply]
Could you give an example of how can write it?

**Reply by dgosbell** [reply]
Sure, if you can post a copy of your code instead of just a screenshot

**Reply by Lomarques-s** [reply]
Of course!

EVALUATE
VAR Share_Context = DIVIDE([Faturado Kg],CALCULATE([Faturado Kg],ALLSELECTED(dClientes)))

VAR vTable = SUMMARIZE(dClientes,dClientes[Cliente],"_Particip%",[Share %])

VAR vCumulative = ADDCOLUMNS(vTable,"Acumulado",SUMX(FILTER(vTable,[_Particip%]<=Share_Context),[Share %]))

RETURN

vCumulative

**Comment by dgosbell** [comment]
Defining it as a measure (which will get re-calculated for each row) would look something like the following

```
DEFINE
MEASURE dClientes[Share_Context] = DIVIDE([Faturado Kg],CALCULATE([Faturado Kg],ALLSELECTED(dClientes)))

EVALUATE

VAR vTable = SUMMARIZE(dClientes,dClientes[Cliente],"_Particip%",[Share %])

VAR vCumulative = ADDCOLUMNS(vTable,"Acumulado",SUMX(FILTER(vTable,[_Particip%]<=[Share_Context]),[Share %]))

RETURN

vCumulative
```

Or since the Share_Content expression is only referenced once you could also just move it inline inside the FILTER() call
```
EVALUATE

VAR vTable = SUMMARIZE(dClientes,dClientes[Cliente],"_Particip%",[Share %])

VAR vCumulative = ADDCOLUMNS(vTable,"Acumulado",SUMX(FILTER(vTable,[_Particip%]<=
CALCULATE(DIVIDE([Faturado Kg],CALCULATE([Faturado Kg],ALLSELECTED(dClientes))))
),[Share %]))

RETURN

vCumulative
```

**Reply by Lomarques-s** [reply]
Thanks for the help.

I did as suggested, I tried both ways, but both continue to return the same result.

see below

1st attempt

![image](https://user-images.githubusercontent.com/121581060/213873908-075d0a16-88fd-4752-9a89-30be1cff535a.png)



2nd attempt
![image](https://user-images.githubusercontent.com/121581060/213873947-f5535a34-f8ac-4fcd-b9d2-bd79287a5151.png)

Remembering that I'm trying to do this in Power Pivot.

I noticed that in Power Pivot it doesn't work, but in Power BI it does. I don't understand why, because it's the same DAX language. Wouldn't that be some kind of bug?

I really need to do a percentage cumulative calculation in Power Pivot. Do you know other alternatives to do this?
My data model contains a sales fact table with 5 million rows and a customer dimension table with 50k records.

**Reply by dgosbell** [reply]
Actually I think because of the ALLSELECTED the measure is still returning the same value for each row. Something like the following should work where we store the percentage value before looping over the table a second time with the FILTER function.

```
EVALUATE

VAR vTable = SUMMARIZE(dClientes,dClientes[Cliente],"_Particip%",[Share %])

VAR vCumulative = ADDCOLUMNS(vTable,"Acumulado",
                    var _current% = [_Particip%]  -- store the % value for the current row
                    return SUMX(FILTER(vTable,[_Particip%]<=_current%) -- filter vTable for all percentages less than the current row's value
                                         ,[_Particip%])) 

RETURN

vCumulative
```

The DAX should work the same in Power BI and Power Pivot, it's just that Power BI has some newer functions. I can only guess that there might be other differences in the data model.

**Reply by Lomarques-s** [reply]
Now it's working. Thanks a lot for the help. But now that I got it, a new problem has arisen, the measurement is taking too long to load the data. I need to think of a way to make the measure more optimized.
